README.md
# lock\_rdlm cookbook

Uses [RDLM](https://github.com/thefab/restful-distributed-lock-manager) to enforce critical sections between Chef clients.  
Also supports managing the RDLM server.  

## Supported Platforms

For clients: Nothing native, so anything goes I guess.  
For lock server: Tested on CentOS 6.7.  

## Attributes

| Key                          | Type    | Used by            | Description                                                                       | Default |
|:-----------------------------|:--------|:-------------------|:----------------------------------------------------------------------------------|:--------|
| `['lock_rdlm']['scheme']`   | String  | Clients            | Default for lock scheme                                                          | http    |
| `['lock_rdlm']['hostname']` | String  | Clients            | Default for lock server name                                                     | nil     |
| `['lock_rdlm']['port']`     | Integer | Server and Clients | Port for lock server. Used as default by clients                                 | 7305    |
| `['lock_rdlm']['wait']`     | Integer | Clients            | Default for number of seconds to wait to acquire the lock before raising an error | 5       |
| `['lock_rdlm']['lifetime']` | Integer | Clients            | Default for how logn the lock hols before expiring by itself                      | 300     |

## Recipes

### lock\_rdlm::server
Installs and configures the simple RDLM server.  
It uses the `['lock_rdlm']['port']` attribute to determine the port the daemon will be listening on.  
I'm using an init file tested on CentOS 6.7.  

## Library resources

### Common parameters
Used to locate the lock server, appear under `additional_config` in method descriptions:

* `lock_url`: How to reach the lock server. Defaults to building the url from attributes (if possible)
* `lock_wait`: Number of seconds to wait to acquire the lock before raising an error. Defaults to `node['lock_rdlm']['wait']`
* `lock_lifetime`: Number of seconds before the lock will expire on its own. Defaults to `node['lock_rdlm']['lifetime']`

### LockRDLM::with\_lock
Yields (runs a code block) while locking the resource specified on the lock.  
Parameters:

* `node`: Calling node's object
* `lock_resource`: name for lock
* "additional_config" detailed above

Can be thought of like the [synchronized](https://docs.oracle.com/javase/tutorial/essential/concurrency/locksync.html) keyword in java.  
Example:

```ruby
# Add apple to databag if it has none
LockRDLM::with_lock(node,'dbvendingjuice') do
  db=data_bag_item('vending','juice')
  db['kinds']<<'apple' unless db['kinds'].member? 'apple'
  db.save
end
```
This will ensure the databag has only one `apple` item, even if several clients run this recipe at the same time.

### LockRDLM::assign\_identity

Used to assign a unique identity.  
Parameters:

* `node`: Calling node's object
* `assignment_path`: Path for attribute to store unique value in
* `range`: Pool of possibe values
* `additional_config`: Detailed above

#### Resulting effects
Upon successful completion of the function, this node will be assigned a unique identity in its node object.  
The identity will be stored as a node attribute specified by `assignment_path`, where `[:a,0,'bla']` is used to address the node attribute `node[:a][0]['bla']`.  
The value is selected from `range` and is guranteed to be different from any value currently present on other nodes in the Chef server.  
Selection is done without concurrency, thanks to the lock.  
The resulting value is saved on both the current `node` object (for use in other parts of the recipe) and in the server's version of the `node` object immediately, to stop other nodes from choosing the same value. This is important to mention because by default, Chef client only updates the server's version of the `node` object if the run is successful. In our case, this attribute is updated during resource compilation.  
There are several ways this method can fail:

* Can't talk to lock server  
    Fix the server, firewalls etc.
* Can't get lock lock (other client is busy with lock).
    * My library might be buggy and not release the lock for some reason. Create a PR/Issue :)
    * Enumerating all of the nodes takes longer than the lock time constraints. Allow more `wait` time.
* Range is exhausted
    * Increase the range
    * Delete dead nodes from the Chef server

#### Basic walk-through
Assuming a working and reachable lock server:

1. Function is called with the following:
    * `assignment_path = [:slasher,:id]`
    * `range = (1..5).to_a`
    * `lock_url = 'http://lock:8080/'`
    * `lock_wait = 5`
    * `lock_lifetime = 300`
2. Current value of attribute is checked (`node[:slasher][:id]`). If not-empty, function returns. Assuming empty
3. Lock is locked for a normalized version of the `assignment_path` (`slasherid`)
4. If lock can't be locked for some reason, raise error
5. All node objects on Chef server are enumerated for that value. Resulting collection is filtered out from range.
6. Take a single element from range (`first`). If empty, raise error (release lock before). Assuming got `3`.
7. Load our node object from Chef server, modify the node attribute and save it.
8. Update the current node's attribute (`node.normal[:slasher][:id]=3`)
9. Release lock
10. Return 3

### LockRDLM::find\_duplicate\_identity
Parameters:

* `node`: Calling node's object
* `assignment_path`: Path for attribute unique value is stored in
* `only_me`: Only check for conflicts involving my node
* `additional_config`: Detailed above

Used to enforce uniqueness of the identity attribute without modifying anything.  
Is useful in monitoring.  

#### Return value
Empty hash (`{}`) if there are no duplicates.  
If there are duplicates, the hash is composed of a key for the duplicate identity and the value is an array of nodes holding said value, like this:

```ruby
{
  3 => ['node1.backslasher.net', 'node2.backslasher.net']
}
```

## License and Authors
Licensed [GPL v2](http://choosealicense.com/licenses/gpl-2.0/)  
Author:: [Nitzan Raz](https://github.com/BackSlasher) ([backslasher](http://backslasher.net/))


CHANGELOG.md
# 0.1.3
- [backslasher] Updated README

# 0.1.2
- [backslasher] mutex to lock

# 0.1.1
- [backslasher] Updated README

# 0.1.0
Initial release of mutex_rdlm


test/cookbooks/lock_rdlm_test/metadata.rb
name             'lock_rdlm_test'
maintainer       'Backslasher'
maintainer_email 'nitz.raz@gmail.com'
license          'GPL v2'
description      'Wrapper cookbook to test lock_rdlm'
long_description 'Wrapper cookbook to test lock_rdlm'
version          '0.1.0'

depends          'lock_rdlm'


test/cookbooks/mutex_rdlm_test/metadata.rb
name             'mutex_rdlm_test'
maintainer       'Backslasher'
maintainer_email 'nitz.raz@gmail.com'
license          'GPL v2'
description      'Wrapper cookbook to test mutex_rdlm'
long_description 'Wrapper cookbook to test mutex_rdlm'
version          '0.1.0'

depends          'mutex_rdlm'


test/cookbooks/mutex_identity_test/metadata.rb
name             'mutex_identity_test'
maintainer       'Backslasher'
maintainer_email 'nitz.raz@gmail.com'
license          'GPL v2'
description      'Wrapper cookbook to test mutex_identity'
long_description 'Wrapper cookbook to test mutex_identity'
version          '0.1.0'

depends          'mutex_identity'


libraries/lock.rb
# cookbook: lock_rdlm
# library: helpers.rb
#
# Author: Nitzan
# GPLv2
#
# Assign a unique value to a node's attribute

module LockRDLM

  # Locks a lock. Returns a lock object (used to release it) if successful
  def self.lock_acquire(server,lock_name,client_name,wait,lifetime)
    require 'net/http'
    uri = URI(server)
    uri.path="/locks/#{lock_name}"
    require 'json'
    data = {title: client_name, wait: wait, lifetime: lifetime }.to_json
    resp = Net::HTTP.new(uri.host,uri.port).post(uri,data) #TODO is it sending ok?
    if resp.code == '201'
      obj = resp['Location']
      return obj
    elsif resp.code == '408'
      raise 'Lock request timed out' #TODO use my exception
    elsif resp.code == '409'
      raise 'Lock request deleted by admin' #TODO use my exception
    elsif resp.code == '400'
      raise 'Bad lock request' #TODO use my exception
    else
      raise "Unknown response #{resp.code}" #TODO use my exception
    end
  end

  def self.lock_check(lock_object)
    require 'net/http'
    require 'json'
    uri = URI(lock_object)
    resp = Net::HTTP.new(uri.host,uri.port).get(uri)
    if resp.code == '404'
      return nil # no such lock
    elsif resp.code == '200'
      return JSON.parse(resp.body)
    else
      raise "Unknown response #{resp.code}" #TODO use my exception
    end
  end

  def self.lock_release(lock_object)
    require 'net/http'
    uri = URI(lock_object)
    resp = Net::HTTP.new(uri.host,uri.port).delete(uri)
    if resp.code == '204'
      return
    elsif resp.code == '404'
      raise 'Lock not found' #TODO use my exception
    else
      raise "Unknown response #{resp.code}"
    end
  end

  def self._normalize_name(source)
    source.gsub(/\W|_/,'')
  end

  def self.with_lock(node,lock_resource, lock_url: nil, lock_wait: nil, lock_lifetime: nil)

    unless lock_url then
      url_blocks = node['lock_rdlm'].values_at(:scheme,:hostname,:port)
      raise 'cannot deduce lock server. Please specify one using attributes or function arguments' if url_blocks.any?{|v|!v} #TODO use my exception
      lock_url = "#{url_blocks[0]}://#{url_blocks[1]}:#{url_blocks[2]}"
    end

    lock_wait||=node['lock_rdlm']['wait']
    lock_lifetime||=node['lock_rdlm']['lifetime']

    # Start working
    lock=lock_acquire(lock_url,
               _normalize_name(lock_resource),
               _normalize_name(node.name),
               lock_wait,lock_lifetime)
    begin
      yield if block_given?
    ensure # Always delete lock
      lock_release(lock)
    end
  end
end


libraries/helpers.rb
# cookbook: mutex_rdlm
# library: helpers.rb
#
# Author: Nitzan
# GPLv2
#
# Helper methods for mutexing

module MutexRDLM
  # get value from node
  def _get_value(node,assignment_path)
    node.attributes.merged_attributes(*assignment_path)
  end

  # Set a value in a vivid hash/array, recursively
  def _vivid_set(vivid,assignment_path,value)
    if assignment_path.length==1
      vivid[assignment_path.first]=value
    else
      _vivid_set(vivid[assignment_path.first],assignment_path.drop(1))
    end
  end

  # set value in node's normal
  def _set_normal(node,assignment_path,value)
    _vivid_set(node.normal,assignment_path,value)
  end

  # set value in node's server
  def _set_server(node,assignment_path,value)
    s_node = Chef::Node.load(node.name)
    _set_normal(s_node,assignment_path,value)
    s.save
  end
end


libraries/assign_identity.rb
# cookbook: lock_rdlm
# library: helpers.rb
#
# Author: Nitzan
# GPLv2
#
# Assign a unique value to a node's attribute

module LockRDLM
  # get value from node
  def self._get_value(node,assignment_path)
    node.attributes.merged_attributes(*assignment_path)
  end

  # Set a value in a vivid hash/array, recursively
  def self._vivid_set(vivid,assignment_path,value)
    if assignment_path.length==1
      vivid[assignment_path.first]=value
    else
      _vivid_set(vivid[assignment_path.first],assignment_path.drop(1),value)
    end
  end

  # set value in node's normal
  def self._set_normal(node,assignment_path,value)
    _vivid_set(node.normal,assignment_path,value)
  end

  # set value in node's server
  def self._set_server(node,assignment_path,value)
    s_node = Chef::Node.load(node.name)
    _set_normal(s_node,assignment_path,value)
    s_node.save
  end

  def self.assign_identity(node,assignment_path,range,*additional_config)
    current_value =  _get_value(node,assignment_path)
    return current_value if current_value # abort if value is already set
    # We have to invent a new value
    with_lock(node,assignment_path.join,*additional_config) do
      # Find existing values
      node_names = Chef::Node.list.keys
      existing_values = node_names.map do |name|
        n = Chef::Node.load(name)
        _get_value(n,assignment_path)
      end
      remaining_range = range - existing_values
      # Choose my value
      if remaining_range.empty?
        raise 'Identity range is exhausted' #TODO use custom exception
      else
        my_value = remaining_range.first
        _set_normal(node,assignment_path,my_value)
        _set_server(node,assignment_path,my_value)
        return my_value
      end
    end
  end

  def self.find_duplicate_identity(node,assignment_path, *additional_config)

    # Params
    only_hash = additional_config.find{|i|i.class==Hash and i[:only_me]}
    if only_hash
      only_me = only_hash[:only_me]
      additional_config.delete(only_hash)
    else
      only_me=false
    end

    with_lock(node,assignment_path.join,*additional_config) do
      node_names = Chef::Node.list.keys
      existing_map = node_names.map do |name|
        n = Chef::Node.load(name)
        val = _get_value(n,assignment_path)
        [name,val]
      end
      grouped = existing_map.
        select{|v|v[1]}. # Filter no-value
        group_by{|v|v[1]}. # group
        map{|k,v|[k,v.map{|i|i[0]}]}.to_h # hash group
      grouped.select!{|k,v|v.include?(node.name)} if only_me
      dups = grouped.select{|k,v|v.count>1}
      return dups
    end
  end
end


libraries/mutex.rb
# cookbook: mutex_rdlm
# library: helpers.rb
#
# Author: Nitzan
# GPLv2
#
# Assign a unique value to a node's attribute

module MutexRDLM

  # Locks a mutex. Returns a mutex object (used to release it) if successful
  def self.mutex_lock(server,mutex_name,client_name,wait,lifetime)
    require 'net/http'
    uri = URI(server)
    uri.path="/locks/#{mutex_name}"
    require 'json'
    data = {title: client_name, wait: wait, lifetime: lifetime }.to_json
    resp = Net::HTTP.new(uri.host,uri.port).post(uri,data) #TODO is it sending ok?
    if resp.code == '201'
      obj = resp['Location']
      return obj
    elsif resp.code == '408'
      raise 'Lock request timed out' #TODO use my exception
    elsif resp.code == '409'
      raise 'Lock request deleted by admin' #TODO use my exception
    elsif resp.code == '400'
      raise 'Bad lock request' #TODO use my exception
    else
      raise "Unknown response #{resp.code}" #TODO use my exception
    end
  end

  def self.mutex_check(mutex_object)
    require 'net/http'
    require 'json'
    uri = URI(mutex_object)
    resp = Net::HTTP.new(uri.host,uri.port).get(uri)
    if resp.code == '404'
      return nil # no such mutex
    elsif resp.code == '200'
      return JSON.parse(resp.body)
    else
      raise "Unknown response #{resp.code}" #TODO use my exception
    end
  end

  def self.mutex_release(mutex_object)
    require 'net/http'
    uri = URI(mutex_object)
    resp = Net::HTTP.new(uri.host,uri.port).delete(uri)
    if resp.code == '204'
      return
    elsif resp.code == '404'
      raise 'Lock not found' #TODO use my exception
    else
      raise "Unknown response #{resp.code}"
    end
  end

  def self._normalize_name(source)
    source.gsub(/\W|_/,'')
  end

  def self.with_mutex(node,mutex_resource, mutex_url: nil, mutex_wait: nil, mutex_lifetime: nil)

    unless mutex_url then
      url_blocks = node['mutex_rdlm'].values_at(:scheme,:hostname,:port)
      raise 'cannot deduce mutex server. Please specify one using attributes or function arguments' if url_blocks.any?{|v|!v} #TODO use my exception
      mutex_url = "#{url_blocks[0]}://#{url_blocks[1]}:#{url_blocks[2]}"
    end

    mutex_wait||=node['mutex_rdlm']['wait']
    mutex_lifetime||=node['mutex_rdlm']['lifetime']

    # Start working
    mutex=mutex_lock(mutex_url,
               _normalize_name(mutex_resource),
               _normalize_name(node.name),
               mutex_wait,mutex_lifetime)
    begin
      yield if block_given?
    ensure # Always delete mutex
      mutex_release(mutex)
    end
  end
end
